cogeo_mosaic/backends/base.py

```diff
@@ -196,35 +228,36 @@ def update(
         **kwargs,
     ):
         """Update existing MosaicJSON on backend."""
+        if self.mode == "r":
+            raise ValueError(
+                "Can only update a mosaic opened in 'r+' or 'w' mode, not r."
+            )
+
+        mosaic = deepcopy(self.mosaic_def)
         new_mosaic = MosaicJSON.from_features(
             features,
-            self.mosaic_def.minzoom,
-            self.mosaic_def.maxzoom,
+            mosaic.minzoom,
+            mosaic.maxzoom,
             quadkey_zoom=self.quadkey_zoom,
             quiet=quiet,
             **kwargs,
         )
 
+        # Update Tiles
         for quadkey, new_assets in new_mosaic.tiles.items():
-            tile = mercantile.quadkey_to_tile(quadkey)
-            assets = self.assets_for_tile(*tile)
+            assets = mosaic.tiles.get(quadkey, [])
             assets = [*new_assets, *assets] if add_first else [*assets, *new_assets]
+            mosaic.tiles[quadkey] = assets
 
-            # add custom sorting algorithm (e.g based on path name)
-            self.mosaic_def.tiles[quadkey] = assets
-
-        bounds = bbox_union(new_mosaic.bounds, self.mosaic_def.bounds)
+        # Update Metadata
+        bounds = bbox_union(new_mosaic.bounds, mosaic.bounds)
 
-        self.mosaic_def._increase_version()
-        self.mosaic_def.bounds = bounds
-        self.mosaic_def.center = (
+        mosaic._increase_version()
+        mosaic.bounds = bounds
+        mosaic.center = (
             (bounds[0] + bounds[2]) / 2,
             (bounds[1] + bounds[3]) / 2,
-            self.mosaic_def.minzoom,
+            mosaic.minzoom,
         )
 
-        # We only write if path is set
-        if self.path:
-            self.write(overwrite=True)
-
-        return
+        self.write(mosaic, overwrite=True)
```

Comment:
here we call `write` which will write the mosaic and set `mosaic_def` to the new value